### PR TITLE
Add background worker

### PR DESCRIPTION
**Pull request by olbjo**
Handles saving to file and reading ethernet status.

### CODE REVIEW

File: src/device/pf_pdport.h [hunk 1]
```diff
@@ -150,6 +150,33 @@ void pf_pdport_start_linkmonitor (pnet_t * net);
...
+/**
+ * Get current ethernet status for local port.
+ * The status is read in the background worker
+ * task.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range: 1 .. num_physical_ports
+ */
+pnal_eth_status_t pf_pdport_get_eth_status (pnet_t * net, int loc_port_num);
```
Comment (pyhys) [inline src/device/pf_pdport.h:178]:
```
* Get current ethernet status for local port.
* This is a non-blocking function reading from mutex-protected memory. It is 
   the responsibility of another thread (elsewhere in the stack) to make sure that the information is up to date.
```

File: src/device/pf_cmina.h [hunk 2]
```diff
@@ -213,6 +213,17 @@ bool pf_cmina_has_timed_out (
...
+/**
+ * Save the IP settings (and station name) to nonvolatile memory if necessary.
+ *
+ * Compares with the content of already stored settings (in order not to
+ * wear out the flash chip)
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param p_ase            In:    Settings to be saved
+ */
+void pf_cmina_save_ase (pnet_t * net, pf_cmina_dcp_ase_t * p_ase);
```
Comment (pyhys) [inline src/device/pf_cmina.h:225]: Is this blocking or not? Does it hand off to background task?

File: src/common/pf_bg_worker.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+static void run_save_file_job (pnet_t * net)
+{
+   LOG_DEBUG (PNET_LOG, "BGW(%d): Running save file job\n", __LINE__);
+   pf_cmina_save_ase (net, &net->cmina_nonvolatile_dcp_ase);
+   pf_fspm_save_im (net);
+   pf_pdport_save_all (net);
+}
+
+static void run_update_ports_status_job (pnet_t * net)
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:114]:
```
/**
 * Run a job updating the status for all physical ports.
 *
 * @param net              InOut: The p-net stack instance
 */
```

File: src/common/pf_bg_worker.c [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            net_handle->pf_bg_worker.events,
+            BG_JOB_EVENT_UPDATE_PORTS_STATUS);
+         {
+            run_update_ports_status_job (net_handle);
+         }
+      }
+   }
+}
+
+static void run_save_file_job (pnet_t * net)
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:106]:
```
/**
 * Run a job saving all files.
 *
 * @param net              InOut: The p-net stack instance
 */
```

File: src/common/pf_bg_worker.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+         "BGW(%d): Unsupported job %" PRIu32 "\n",
+         __LINE__,
+         job_id);
+      return -1;
+   }
+
+   return 0;
+}
+
+static void bg_worker_task (void * arg)
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:70]:
```
/**
 * Task for running in a background thread
 *
 * @param arg              InOut: Thread argument, should be pnet_t *
 */
```

File: src/device/pf_fspm.h [hunk 6]
```diff
@@ -375,6 +375,16 @@ void pf_fspm_save_im_location (pnet_t * net, const char * location);
...
+/**
+ * Save the I&M settings to nonvolatile memory, if necessary.
+ *
+ * Compares with the content of already stored settings (in order not to
+ * wear out the flash chip)
+ *
+ * @param net              InOut: The p-net stack instance
+ */
+void pf_fspm_save_im (pnet_t * net);
+
```
Comment (pyhys) [inline src/device/pf_fspm.h:387]: Is this blocking or not?

File: src/device/pf_pdport.c [hunk 7]
```diff
@@ -1368,3 +1380,53 @@ void pf_pdport_peer_indication (pnet_t * net, int loc_port_num)
...
+      loc_port_num = pf_port_get_next (&port_iterator);
+   }
+}
+
+pnal_eth_status_t pf_pdport_get_eth_status (pnet_t * net, int loc_port_num)
+{
+   pf_port_t * p_port_data;
+   pnal_eth_status_t eth_status;
+
+   p_port_data = pf_port_get_state (net, loc_port_num);
```
Comment (pyhys) [inline src/device/pf_pdport.c:1425]: This can be part of first line.

**Review by pyhys — COMMENTED** [review]
Great work!
Should the flags be more fine-grained, so we do not need to save all files at once?

File: src/common/pf_lldp.c [hunk 8]
```diff
@@ -895,18 +895,8 @@ void pf_lldp_get_link_status (
...
-   if (pnal_eth_get_status (p_port_data->netif.name, &status) != 0)
-   {
-      LOG_ERROR (
-         PF_LLDP_LOG,
-         "LLDP(%d): Failed to read Ethernet port status for port %d\n",
-         __LINE__,
-         loc_port_num);
-      memset (&status, 0, sizeof (status));
-   }
+   status = pf_pdport_get_eth_status (net, loc_port_num);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:899]: OK

File: src/device/pf_cmina.c [hunk 9]
```diff
@@ -353,8 +353,7 @@ int pf_cmina_set_default_cfg (pnet_t * net, uint16_t reset_mode)
       }
       net->cmina_nonvolatile_dcp_ase.product_name[ix] = '\0';
 
-      /* Save to file */
-      pf_cmina_save_ase (net, &net->cmina_nonvolatile_dcp_ase);
+      pf_bg_worker_run_job (net, BG_JOB_SAVE_NVM_DATA);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_cmina.c:356]: OK

File: src/device/pf_pdport.h [hunk 10]
```diff
@@ -150,6 +150,33 @@ void pf_pdport_start_linkmonitor (pnet_t * net);
  */
 void pf_pdport_periodic (pnet_t * net);
 
+/**
+ * Save PDPort configuration to NVM for all ports.
+ * This function is intended to be executed in the background task.
+ *
+ * @param net              InOut: The p-net stack instance
+ */
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/device/pf_pdport.h:158]: OK

File: src/common/pf_bg_worker.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Initialize the background worker.
+ *
+ * @param net              InOut: The p-net stack instance
+ */
+void pf_bg_worker_init (pnet_t * net);
+
+/**
+ * Start a background job
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_bg_worker.h:40]: OK

File: src/common/pf_bg_worker.c [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+   LOG_DEBUG (PNET_LOG, "BGW(%d): Running save file job\n", __LINE__);
+   pf_cmina_save_ase (net, &net->cmina_nonvolatile_dcp_ase);
+   pf_fspm_save_im (net);
+   pf_pdport_save_all (net);
+}
+
+static void run_update_ports_status_job (pnet_t * net)
+{
+   pf_pdport_update_eth_status (net);
+}
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_bg_worker.c:117]: OK

File: src/common/pf_bg_worker.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+ * @param net              InOut: The p-net stack instance
+ * @param job_id           In:    Job to run
+ * @return  0  if operation is successfully initiated
+ *          -1 if an error occurred.
+ */
+int pf_bg_worker_run_job (pnet_t * net, pf_bg_job_t job_id);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_bg_worker.h:46]: Changed to pf_bg_worker_start_job()

OK

File: src/common/pf_bg_worker.c [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      }
+   }
+}
+
+static void run_save_file_job (pnet_t * net)
+{
+   LOG_DEBUG (PNET_LOG, "BGW(%d): Running save file job\n", __LINE__);
+   pf_cmina_save_ase (net, &net->cmina_nonvolatile_dcp_ase);
+   pf_fspm_save_im (net);
+   pf_pdport_save_all (net);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_bg_worker.c:111]: Added (void)
All the save operations internally log error (and debug logs)
The pf_pdport_save_all() has a handles a special case during startup when a mutex is not yet created.
OK

File: src/device/pf_pdport.h (hunk 1 above)
Comment (olbjo) [reply src/device/pf_pdport.h:178]: OK
Updated description, added @return

File: src/device/pf_cmina.h (hunk 2 above)
Comment (olbjo) [reply src/device/pf_cmina.h:225]: OK

File: src/common/pf_bg_worker.c (hunk 3 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:114]: Removed the inernal functions and instead calling the functions doing the job directly in the event handler loop. I think this is a better solution after splitting up the save job to different jobs (IM,ASE,PDPORT).
OK

File: src/common/pf_bg_worker.c (hunk 4 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:106]: OK, se previous comment.

File: src/common/pf_bg_worker.c (hunk 5 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:70]: OK

File: src/device/pf_fspm.h (hunk 6 above)
Comment (olbjo) [reply src/device/pf_fspm.h:387]: OK

**Comment by olbjo** [comment]
I also split the file saving into separate jobs

File: src/common/pf_bg_worker.c (hunk 5 above)
Comment (pyhys) [reply src/common/pf_bg_worker.c:82]: I think there is still some static function without docstring.

File: src/common/pf_bg_worker.c [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+int pf_bg_worker_start_job (pnet_t * net, pf_bg_job_t job_id)
+{
+   if (net == NULL)
+   {
+      return -1;
+   }
+
+   switch (job_id)
+   {
+
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:57]: Make sure to run clang-format. Should there be an empty line here?

File: src/common/pf_bg_worker.c [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+            pf_fspm_save_im (net);
+         }
+      }
+      if (flags & BG_JOB_EVENT_SAVE_PDPORT_NVM_DATA)
+      {
+         os_event_clr (
+            net->pf_bg_worker.events,
+            BG_JOB_EVENT_SAVE_PDPORT_NVM_DATA);
+         {
+            pf_pdport_save_all (net);
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:114]: `pf_pdport_save_all()` returns an int, so we need an `(void)`

File: src/common/pf_bg_worker.c [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+   uint32_t flags = 0;
+
+   for (;;)
+   {
+      os_event_wait (net->pf_bg_worker.events, mask, &flags, OS_WAIT_FOREVER);
+
+      if (flags & BG_JOB_EVENT_SAVE_ASE_NVM_DATA)
+      {
+         os_event_clr (net->pf_bg_worker.events, BG_JOB_EVENT_SAVE_ASE_NVM_DATA);
+         {
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:97]: Do we need this level of block? Remove additional `{`

File: src/common/pf_bg_worker.h [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Background worker jobs
+ */
+typedef enum pf_bg_job
+{
+   BG_JOB_UPDATE_PORTS_STATUS = 0, /**< Update status for all ports  */
```
Comment (pyhys) [inline src/common/pf_bg_worker.h:28]: Also enum content needs to start with PF. Maybe
`PF_BGJOB_UPDATE_PORTS_STATUS`

File: src/device/pf_cmina.h [hunk 19]
```diff
@@ -213,6 +213,19 @@ bool pf_cmina_has_timed_out (
    uint16_t period,
    uint16_t factor);
 
+/**
+ * Save the IP settings (and station name) to nonvolatile memory if necessary.
+ * This is a blocking call intended to be executed by the background worker.
+ *
+ *
```
Comment (pyhys) [inline src/device/pf_cmina.h:220]: Remove one blank line

File: src/device/pf_pdport.c [hunk 20]
```diff
@@ -970,29 +988,21 @@ static void pf_pdport_handle_link_up (pnet_t * net, int loc_port_num)
  */
 static void pf_pdport_monitor_link (pnet_t * net, int loc_port_num)
 {
-   pnal_eth_status_t status;
+   pnal_eth_status_t eth_status;
```
Comment (pyhys) [inline src/device/pf_pdport.c:991]: ` pnal_eth_status_t eth_status = pf_pdport_get_eth_status (net, loc_port_num);`

File: test/test_lldp.cpp [hunk 21]
```diff
@@ -835,6 +835,9 @@ TEST_F (LldpTest, LldpGetLinkStatus)
       PNAL_ETH_MAU_COPPER_100BaseTX_FULL_DUPLEX;
    memset (&link_status, 0xff, sizeof (link_status));
 
+   /* The ethernet status is update by the background worker */
```
Comment (pyhys) [inline test/test_lldp.cpp:838]: `is updated`

File: src/common/pf_bg_worker.c [hunk 22]
```diff
@@ -0,0 +1,127 @@
+
+/*********************************************************************
+ *        _       _         _
+ *  _ __ | |_  _ | |  __ _ | |__   ___
+ * | '__|| __|(_)| | / _` || '_ \ / __|
+ * | |   | |_  _ | || (_| || |_) |\__ \
+ * |_|    \__|(_)|_| \__,_||_.__/ |___/
+ *
+ * www.rt-labs.com
+ * Copyright 2032 rt-labs AB, Sweden.
```
Comment (pyhys) [inline src/common/pf_bg_worker.c:10]: 2021

**Review by pyhys — COMMENTED** [review]
Sweet. 
Minor comments only,

File: src/common/pf_bg_worker.c (hunk 22 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:10]: 😄

File: src/common/pf_bg_worker.c (hunk 15 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:57]: OK, removed.
clang-format did not remove is automatically

File: src/common/pf_bg_worker.c (hunk 16 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:114]: OK

File: src/common/pf_bg_worker.c (hunk 17 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:97]: OK

File: src/common/pf_bg_worker.h (hunk 18 above)
Comment (olbjo) [reply src/common/pf_bg_worker.h:28]: OK

File: src/device/pf_pdport.c (hunk 20 above)
Comment (olbjo) [reply src/device/pf_pdport.c:991]: OK

File: test/test_lldp.cpp (hunk 21 above)
Comment (olbjo) [reply test/test_lldp.cpp:838]: OK

File: src/common/pf_bg_worker.c (hunk 22 above)
Comment (olbjo) [reply src/common/pf_bg_worker.c:10]: OK

File: src/device/pf_pdport.h [hunk 23]
```diff
@@ -150,6 +150,33 @@ void pf_pdport_start_linkmonitor (pnet_t * net);
...
+int pf_pdport_save_all (pnet_t * net);
+
+/**
+ * Update the ethernet status for all ports.
+ * This function is intended to be executed in the background task.
+ *
+ * @param net              InOut: The p-net stack instance
+ */
+void pf_pdport_update_eth_status (pnet_t * net);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_pdport.h:171]: Now updated

File: src/common/pf_bg_worker.c (hunk 5 above)
Comment (pyhys) [reply src/common/pf_bg_worker.c:81]: Now updated

File: src/device/pf_pdport.c (hunk 7 above)
Comment (pyhys) [reply src/device/pf_pdport.c]: Updated

**Review by pyhys — APPROVED** [review]
Excellent!

**Comment by pyhys** [comment]
All tests pass in ART Tester

**Pull request merged**